File: C/3/5.c
#include <stdio.h>
#include <string.h>
#include <limits.h>

#define LIM 99999

/* reverse:  reverse string s in place */
void reverse(char s[])
{
    int c, i, j;

    for (i = 0, j = strlen(s)-1; i < j; i++, j--) {
        c = s[i];
        s[i] = s[j];
        s[j] = c;
    }
}

/**
 * Converts integer x to a string with the number converted to base b.
 *
 * @param int    x The integer to be converted
 * @param char[] s The resultant string
 * @param int    b The base to be changed to
 *
 * @return void
 */
void itob(int x, char s[], short int b)
{
    /* Set of possible characters */
    char characters[] = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    int i = 0;
    /* Handling int overflow */
    long n = x;

    if (b < 0 || b > 36) {
        s[0] = '\0';
        printf("Invalid radix: %hi!\n", b);
        return;
    }

    if (n < 0)
        n = -n;

    do {
        /**
         * The remainder of the division of and and the base is a number in base
         * 10 (because n is base 10). We use it as the position in our set of
         * possible characters.
         */
        s[i++] = characters[n % b];
    } while((n /= b) > 0);
    s[i] = '\0';

    reverse(s);
}

main()
{
    char result[LIM];

    itob(10, result, 2);
    printf("itob(10, 2) => '%s'\n", result);

    itob(15, result, 16);
    printf("itob(15, 16) => '%s'\n", result);

    itob(32, result, 2);
    printf("itob(32, 2) => '%s'\n", result);

    itob(2130440, result, 10);
    printf("itob(2130440, 10) => '%s'\n", result);

    itob(128, result, 16);
    printf("itob(128, 16) => '%s'\n", result);

    itob(INT_MAX, result, 2);
    printf("itob(INT_MAX, 2) => '%s'\n", result);

    itob(INT_MIN, result, 2);
    printf("itob(INT_MIN, 2) => '%s'\n", result);

    itob(INT_MIN, result, 16);
    printf("itob(INT_MIN, 16) => '%s'\n", result);
}

File: rustlang/3/5.rs
/// Converts integer x to a string.
///
/// @param  i32 x The integer to be converted
/// @param  i16 b The base to be changed to
///
/// @return String  The resultant string
fn itob(x: i32, b: i16) -> String
{
    // Set of possible characters
    let characters = String::from("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ");

    let mut result = String::new();

    if b < 0 || b > 36 {
        print!("Invalid radix: {}!\n", b);
        return result;
    }

    // Handling overflow
    let mut n: i64 = x as i64;
    if n < 0 {
        n = -n;
    }

    loop {
        // get next digit
        let c = characters.chars().nth((n % (b as i64)) as usize).unwrap();
        result.push(c);

        n /= b as i64;
        if n <= 0 {
            break;
        }
    }

    result.chars().rev().collect::<String>()
}

fn main()
{
    print!("itob(10, 2) => '{}'\n", itob(10, 2));

    print!("itob(15, 16) => '{}'\n", itob(15, 16));

    print!("itob(32, 2) => '{}'\n", itob(32, 2));

    print!("itob(2130440, 10) => '{}'\n", itob(2130440, 10));

    print!("itob(128, 16) => '{}'\n", itob(128, 16));

    print!("itob(std::i32::MAX, 2) => '{}'\n", itob(std::i32::MAX, 2));

    print!("itob(std::i32::MIN, 2) => '{}'\n", itob(std::i32::MIN, 2));

    print!("itob(std::i32::MIN, 16) => '{}'\n", itob(std::i32::MIN, 16));
}
